packages/next/src/build/transpile-config.ts

```diff
@@ -0,0 +1,235 @@
+    target: ['node', 'es2020'],
+    resolve: {
+      modules: ['node_modules'],
+      extensions: ['.ts', '.mts', '.cts', '.js', '.mjs', '.cjs'],
+      // Need to resolve @swc/helpers/_/ alias for next config as async function:
+      // Module not found: Can't resolve '@swc/helpers/_/_async_to_generator'
+      alias: {
+        '@swc/helpers/_': join(
+          dirname(require.resolve('@swc/helpers/package.json')),
+          '_'
+        ),
+      },
+      plugins: [
+        new JsConfigPathsPlugin(
+          tsConfig.compilerOptions?.paths ?? {},
+          resolvedBaseUrl
+        ),
+      ],
+    },
+    plugins: [new ProfilingPlugin({ runWebpackSpan, rootDir: cwd })],
+    module: {
+      rules: [
+        {
+          test: /\.(ts|mts)$/,
+          exclude: /node_modules/,
+          loader: require.resolve('./webpack/loaders/next-swc-loader'),
+          options: {
+            rootDir: cwd,
+            isServer: false,
+            hasReactRefresh: false,
+            // Seems like no need to pass nextConfig to SWC loader
+            nextConfig: {},
+            jsConfig: tsConfig,
+            swcCacheDir: join(distDir, 'cache', 'swc'),
+            supportedBrowsers: undefined,
+          } satisfies SWCLoaderOptions,
+        },
+      ],
+    },
+  }
+
+  // Support tsconfig baseUrl
+  // Only add the baseUrl if it's explicitly set in tsconfig
+  if (!resolvedBaseUrl.isImplicit) {
+    webpackConfig.resolve!.modules!.push(resolvedBaseUrl.baseUrl)
+  }
+
+  const [{ errors }] = await runCompiler(webpackConfig, {
+    runWebpackSpan,
+  })
+
+  if (errors.length > 0) {
+    throw new Error(errors[0].message)
+  }
+}
+
+async function _transpileOnly({
+  nextConfigPath,
+  distDir,
+}: {
+  nextConfigPath: string
+  distDir: string
+}): Promise<NextConfig | undefined> {
+  // Transpile by SWC to check if the config has `import` or `require`.
+  const nextConfig = await readFile(nextConfigPath, 'utf8')
+  const { code } = await transform(nextConfig, swcOptions)
+
+  // Since the code is transpiled to CJS, we only need to check for require.
+  // SWC will also drop types and unused imports.
+  const hasNoImportOrRequire = !code.includes('require(')
+
+  // Transpile-only if there's no import or require.
+  // This will be the most common case and will avoid the need to run bundle.
+  if (hasNoImportOrRequire) {
+    const nextCompiledConfigPath = join(distDir, `next.compiled.config.cjs`)
+
+    await mkdir(dirname(nextCompiledConfigPath), { recursive: true })
+    await writeFile(nextCompiledConfigPath, code)
+
+    return await import(nextCompiledConfigPath)
+  }
+}
+
+export async function transpileConfig({
+  isProd,
+  nextConfigPath,
+  nextConfigName,
+  cwd,
+}: {
+  isProd: boolean
+  nextConfigPath: string
+  nextConfigName: string
+  cwd: string
+}): Promise<NextConfig> {
+  // Since .next will be gitignored, it is OK to use it although distDir might be set on nextConfig.
+  const distDir = join(cwd, '.next')
```

Comment:
Crazy idea here but what if we keep the transpiled source in memory and then we import it via a data URL e.g.

```js
import(`data:text/javascript,${transpiledSource}`)
```

This could help us avoid the need to worry about outputting to a dist dir before we know the dist dir and also could help with perf a bit since we aren't writing to disk. We will need to ensure the transpile source makes anything relative absolute though to avoid resolution issues.

Reply:
Thank you for sharing a brilliant ‼️ idea, and I will investigate it further based on it!
I did consider it before but moved on because of two reasons:

1. Requires much handling edge cases resolving the import path and its extensions
2. Cannot resolve imported file's CJS syntax

Reply:
If we compile to CJS default output target and have automatic ESM support it should just work I believe. Our default output target for webpack bundles in the main bundling is similar.